integrationtests/src/test/java/org/axonframework/integrationtests/eventhandling/TrackingEventProcessorTest.java

```diff
@@ -224,6 +251,44 @@ void tearDown() {
         eventBus.shutDown();
     }
 
+    @Test
+    void testSequenceEventStorageReceivesEachEventOnlyOnce() throws Exception {
+        InMemoryEventStorageEngine historic = new InMemoryEventStorageEngine();
+        InMemoryEventStorageEngine active = new InMemoryEventStorageEngine();
+        SequenceEventStorageEngine sequenceEventStorageEngine = new SequenceEventStorageEngine(historic, active);
+
+        EmbeddedEventStore sequenceEventBus = EmbeddedEventStore.builder().storageEngine(sequenceEventStorageEngine).build();
+
+        initProcessor(TrackingEventProcessorConfiguration.forSingleThreadedProcessing()
+                                                         .andEventAvailabilityTimeout(100, TimeUnit.MILLISECONDS),
+                      b -> {
+                          b.messageSource(sequenceEventBus);
+                          return b;
+                      });
+
+        historic.appendEvents(createEvent(AGGREGATE, 1L, "message1"), createEvent(AGGREGATE, 2L, "message2"));
+        // to make sure tracking tokens match, we need to append the same number of events in the active store
```

Comment:
I don't understand this assumption. Isn't the point of the historic storage that some events will come from the historic, and some. events from the active store? At least this is the case in our production system.

Reply:
That's absolutely correct. In practice, the events are usually in either historic, or the active storage, and only occasionally/temporarily in both. 
However, the tokens returned from the storage engines must continue from one into the next. That means your historic storage will take the globalIndex from the events that the activeStorage has assigned to it. The in-memory storage doesn't support that, so we need to add some 'placeholder' events to make sure the index matches.

Reply:
I did a bit of digging here, and now understand what happens. The first two events from the active storage are not read in playback. Maybe it would be better as documentation to have these two with a different payload and check that only the expected messages are received in the handler?